base/linalg/dense.jl

```diff
@@ -29,6 +29,21 @@ function scale!(X::Array{T}, s::Real) where T<:BlasComplex
     X
 end
 
+function isone(x::StridedMatrix)
+    m, n = size(x)
+    m != n && return false # only square matrices can satisfy x == one(x)
+    for i in 1:m
+        for j in 1:m
```

Comment:
You should be able to just iterate through one of the triangles of the matrix and do two symmetric checks, e.g.
```julia
@inbounds for i = 1:m, j = i:m
    if i == j
        isone(x[i,i]) || return false
    else
        iszero(x[i,j]) || iszero(x[j,i]) || return false
    end
end